pkg/sr/serde.go

```diff
@@ -415,34 +415,61 @@ func (s *Serde) DecodeNew(b []byte) (any, error) {
 	return v, t.decode(b, v)
 }
 
-func (s *Serde) decodeFind(b []byte) ([]byte, tserde, error) {
-	if len(b) < 5 || b[0] != 0 {
-		return nil, tserde{}, ErrBadHeader
+func (s *Serde) decodeFind(b []byte, opts []EncodingOpt) ([]byte, tserde, error) {
+	optsToApply, idopt, indexopt, err := s.filterIDAndIndexOpt(opts)
+	if err != nil {
+		return nil, tserde{}, err
 	}
-	id := binary.BigEndian.Uint32(b[1:5])
-	b = b[5:]
 
-	t := s.loadIDs()[int(id)]
-	if len(t.subindex) > 0 {
-		r := bReader{b}
-		br := io.ByteReader(&r)
-		l, err := binary.ReadVarint(br)
-		if l == 0 { // length 0 is a shortcut for length 1, index 0
-			t = t.subindex[0]
+	var t tserde
+	if idopt != nil {
+		// Load tserde based on the supplied ID.
+		t = s.loadIDs()[int(idopt.ID())]
```

Comment:
The last commit adds the ability to supply `EncodingOpt` to `Decode`, simply to have parity with `Encode`, but I'm not convinced if it makes sense. I included it to run it by you, but I can revert the change if needed.

This is how it works. If you supply options `ID` and `Index` to `Decode` it will find `tserde` based solely on the supplied options, it will not decode the header. In fact, it will assume there is _no header_ and that it's dealing only with the raw encoded bytes. What I don't like about this is that `Serde` can't encode bytes without a header unless you specify your own `SerdeHeader` implementation which skips prepending the header. It would allow you to track the ID and index separately from the encoded payload, although I'm not sure such a use case exists.